source/iterator/iterator.go

```diff
@@ -0,0 +1,254 @@
+	return iter.hasNext(ctx)
+}
+
+// Next returns the next record.
+func (iter *Iterator) Next(_ context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := iter.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := columntypes.TransformRow(row, iter.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[iter.orderingColumn]; !ok {
+		return sdk.Record{}, fmt.Errorf("ordering column %q not found", iter.orderingColumn)
+	}
+
+	key := make(sdk.StructuredData)
+	for i := range iter.keyColumns {
+		val, ok := transformedRow[iter.keyColumns[i]]
+		if !ok {
+			return sdk.Record{}, fmt.Errorf("key column %q not found", iter.keyColumns[i])
+		}
+
+		key[iter.keyColumns[i]] = val
+	}
+
+	rowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	// set a new position into the variable,
+	// to avoid saving position into the struct until we marshal the position
+	positionBytes, err := json.Marshal(transformedRow[iter.orderingColumn])
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal position: %w", err)
+	}
+
+	iter.lastProcessedVal = transformedRow[iter.orderingColumn]
+
+	metadata := sdk.Metadata{
+		metadataTable: iter.table,
+	}
+	metadata.SetCreatedAt(time.Now().UTC())
+
+	return sdk.Util.Source.NewRecordCreate(
+		positionBytes,
+		metadata,
+		key,
+		sdk.RawData(rowBytes),
+	), nil
+}
+
+// Stop stops iterators and closes database connection.
+func (iter *Iterator) Stop() error {
+	if iter.rows != nil {
+		if err := iter.rows.Close(); err != nil {
+			return fmt.Errorf("stop iterator: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// hasNext returns a bool indicating whether the source has the next record to return or not.
+func (iter *Iterator) hasNext(ctx context.Context) (bool, error) {
+	if iter.rows != nil && iter.rows.Next() {
+		return true, nil
+	}
+
+	if err := iter.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	return iter.rows.Next(), nil
+}
+
+// loadRows selects a batch of rows from a database, based on the
+// table, columns, orderingColumn, batchSize and the current position.
+func (iter *Iterator) loadRows(ctx context.Context) error {
+	whereClause := ""
+	args := make([]any, 0, 1)
+	if iter.lastProcessedVal != nil {
+		whereClause = fmt.Sprintf(whereClauseFmt, iter.orderingColumn)
+		args = append(args, iter.lastProcessedVal)
+	}
+
+	query := fmt.Sprintf(querySelectRowsFmt, iter.table, whereClause, iter.orderingColumn, iter.batchSize)
+
+	rows, err := iter.db.QueryxContext(ctx, query, args...)
+	if err != nil {
+		return fmt.Errorf("execute select query %q, %v: %w", query, args, err)
+	}
+
+	iter.rows = rows
+
+	return nil
+}
+
+// populateKeyColumns populates keyColumn from the database's metadata
+// or from the orderingColumn configuration field.
```

Comment:
could you please add additional information about priority choosing keys

Reply:
I've added it to the method's description.